core/src/types.rs

```diff
@@ -609,6 +610,77 @@ impl Mechanism {
+            Self::P256Prehashed => panic!("missing mechanism: P256Prehashed"),
+            #[cfg(feature = "p384")]
+            Self::P384 => panic!("missing mechanism: P384"),
+            #[cfg(feature = "p384")]
+            Self::P384Prehashed => panic!("missing mechanism: P384Prehashed"),
+            #[cfg(feature = "p521")]
+            Self::P521 => panic!("missing mechanism: P521"),
+            #[cfg(feature = "p521")]
+            Self::P521Prehashed => panic!("missing mechanism: P521Prehashed"),
+            #[cfg(feature = "brainpoolp256r1")]
+            Self::BrainpoolP256R1 => panic!("missing mechanism: BrainpoolP256R1"),
+            #[cfg(feature = "brainpoolp256r1")]
+            Self::BrainpoolP256R1Prehashed => panic!("missing mechanism: BrainpoolP256R1Prehashed"),
+            #[cfg(feature = "brainpoolp384r1")]
+            Self::BrainpoolP384R1 => panic!("missing mechanism: BrainpoolP384R1"),
+            #[cfg(feature = "brainpoolp384r1")]
+            Self::BrainpoolP384R1Prehashed => panic!("missing mechanism: BrainpoolP384R1Prehashed"),
+            #[cfg(feature = "brainpoolp512r1")]
+            Self::BrainpoolP512R1 => panic!("missing mechanism: BrainpoolP512R1"),
+            #[cfg(feature = "brainpoolp512r1")]
+            Self::BrainpoolP512R1Prehashed => panic!("missing mechanism: BrainpoolP512R1"),
+            #[cfg(feature = "secp256k1")]
+            Self::Secp256k1 => panic!("missing mechanism: Secp256k1"),
+            #[cfg(feature = "secp256k1")]
+            Self::Secp256k1Prehashed => panic!("missing mechanism: Secp256k1Prehashed"),
+            #[cfg(feature = "sha256")]
+            Self::Sha256 => panic!("missing mechanism: Sha256"),
+            #[cfg(feature = "tdes")]
+            Self::Tdes => panic!("missing mechanism: Tdes"),
+            #[cfg(feature = "totp")]
+            Self::Totp => panic!("missing mechanism: Totp"),
+            #[cfg(feature = "trng")]
+            Self::Trng => panic!("missing mechanism: Trng"),
+            #[cfg(feature = "x255")]
+            Self::X255 => panic!("missing mechanism: X255"),
+            #[cfg(feature = "shared-secret")]
+            Self::SharedSecret => panic!("missing mechanism: SharedSecret"),
+            #[cfg(feature = "rsa2048")]
+            Self::Rsa2048Raw => panic!("missing mechanism: Rsa2048Raw"),
+            #[cfg(feature = "rsa3072")]
+            Self::Rsa3072Raw => panic!("missing mechanism: Rsa3072Raw"),
+            #[cfg(feature = "rsa4096")]
+            Self::Rsa4096Raw => panic!("missing mechanism: Rsa4096Raw"),
+            #[cfg(feature = "rsa2048")]
+            Self::Rsa2048Pkcs1v15 => panic!("missing mechanism: Rsa2048Pkcs1v15"),
+            #[cfg(feature = "rsa3072")]
+            Self::Rsa3072Pkcs1v15 => panic!("missing mechanism: Rsa3096Pkcs1v15"),
+            #[cfg(feature = "rsa4096")]
+            Self::Rsa4096Pkcs1v15 => panic!("missing mechanism: Rsa4096Pkcs1v15"),
+        }
+    }
```

Comment:
What is that for?

Reply:
So that we can give an indication which mechanism is missing, see: https://github.com/Nitrokey/nitrokey-3-firmware/blob/e6132a0b099f8c2b1b575b9c1f1252d112e1d908/components/apps/src/lib.rs#L480

Due to the lack of const formatting, we could otherwise just indicate that there is *some* mechanism missing which can be annoying to debug.  Or am I missing some other trick to get a better error message?

Reply:
Oh, I see.

Can't we instead not have a bunch of const asserts?

Something like:

```rust
const _: () = {
    #[cfg(feature = "se050")]
    assert!(const_contains(mechanisms::ENABLED, Mechanism::BrainpoolP256R1)),
    …
};
```

Where `const_contains` is a `const` function that checks that the mechanism in second argument is contained in the first argument.

AFAIK the compile-time error message should show the line of which assert failed.

Reply:
No, not really.  The check is the other way round.  I don’t want to check the presence of an expected mechanism, I want to check that there is no unexpected mechanism that is enabled but not implemented by a backend or Trussed itself.

Consider the following scenario:  We start to use the new `HmacSha512` mechanism in an application which is currently not enabled in the runner.  This would lead to a runtime error.  So the idea of the check is to list all mechanisms that are activated in `trussed-core` and can be used by applications (`trussed-core::types::Mechanism::ENABLED`) and compare them with the mechanisms that are enabled in `trussed` itself (`trussed::types::IMPLEMENTED_MECHANISMS`).  The list of mechanisms that you see in the linked code are the mechanisms that don’t need to be implemented by `trussed` because they are implemented by other backends, so it’s a whitelist and not the list of checked mechanisms.